### PR TITLE
Add UnitTestTranscriptLogger

### PR DESCRIPTION
**Pull request by tomlm**
Implementation of https://github.com/Microsoft/BotBuilder/issues/5279
* Add TestContext to all unit tests
* Add TranscriptLogger(FileTranscriptogger) to all unit tests

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+                if (this.unitTestMode == true && !started.Contains(transcriptFile))
+                {
+                    started.Add(transcriptFile);
+                    File.Delete(transcriptFile);
+                }
+
+                if (File.Exists(transcriptFile))
+                {
+                    transcript = JsonConvert.DeserializeObject<List<Activity>>(File.ReadAllText(transcriptFile));
```
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: This is by design.  The goal is to create a transcript file for a unit test conversation.  It is not to be efficient file storage for long running conversations.  That was accomplished in the FileTranscriptStore class which you deleted 3 months ago.

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                }
+
```
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: don't care.

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    public class FileTranscriptLogger : ITranscriptLogger
+    {
+
+        private static JsonSerializerSettings jsonSettings = new JsonSerializerSettings()
+        {
+            Formatting = Formatting.Indented,
+            NullValueHandling = NullValueHandling.Ignore,
+        };
+
+        private string folder;
```
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: except for mine, which don't, except when they do

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task LogActivityAsync(IActivity activity)
+        {
+            if (activity != null)
+            {
+                string transcriptFile = Path.Combine(folder, activity.Conversation.Id + ".transcript");
+
+                List<Activity> transcript = null;
+
+                if (this.unitTestMode == true && !started.Contains(transcriptFile))
```
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: unit tests each have their own conversation id, so separate files.  they run sequential, which means they are all operating against their conversation file.

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                {
+                    System.Diagnostics.Trace.TraceInformation($"{activity.From.Name ?? activity.From.Id ?? activity.From.Role}: {((Activity)activity).Text}");
+                }
+                else
+                {
+                    System.Diagnostics.Trace.TraceInformation($"{activity.From.Name ?? activity.From.Id ?? activity.From.Role} [{activity.Type}]");
+                }
+
+                transcript.Add((Activity)activity);
+                File.WriteAllText(transcriptFile, JsonConvert.SerializeObject(transcript, jsonSettings));
```
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: this is for a unit test, by design.

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: added lock

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+namespace Microsoft.Bot.Builder
+{
+    /// <summary>
+    /// FileTranscriptLogger which creates a .transcript file for each conversationId
+    /// </summary>
+    /// <remarks>
+    /// This is a useful class for unit tests
+    /// </remarks>
+    public class FileTranscriptLogger : ITranscriptLogger
```
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: This is by design.  
Activity serialization is built to be future compatible.  (Aka, we are able to round trip properties we don't know about).
This is for unit tests NOT production logging.  
For robust file based see FileTranscriptStore class which you deleted last year.  That stored each activity as a seperate file and then had ITranscriptStore API implementation to build a transcript from the multiple files.  That is NOT the intent of this class.

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: Again, not a real case, but I added lock around file access.  This is for unit tests, not production code.

File: libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder/FileTranscriptLogger.cs]: @tomlm, this is NOT for a unit test. :)

I know that, as this file is public, and it's in our core library. There's no indications developers sholdn't use it for... anything they want.

**Comment by cleemullins** [comment]
@gabog I'm closing this PR out until we've got a chance to close on "Where are we going with Testing?". I spoke with @tomlm about this on Friday, and he's on-board.

**Pull request closed without merging**